### PR TITLE
APIS-4662 - Locking down controllers endpoints

### PR DESCRIPTION
**Pull request by peteslater-ee**

### CODE REVIEW

File: test/controllers/ManageTeamSpec.scala [hunk 1]
```diff
@@ -52,9 +54,9 @@ class ManageTeamSpec extends BaseControllerSpec with SubscriptionTestHelperSugar
 
   val loggedInUser = DeveloperSession(session)
 
-  val tokens = ApplicationToken("clientId", Seq(aClientSecret(), aClientSecret()), "token")
+  val tokens2 = ApplicationToken("clientId", Seq(aClientSecret(), aClientSecret()), "token")
```
Comment (adampridmore) [inline test/controllers/ManageTeamSpec.scala:57]: tokens2?

File: app/controllers/Controller.scala [hunk 2]
```diff
@@ -100,30 +106,34 @@ abstract class ApplicationController()
...
 
-  def permissionThenCapabilityAction(permissions: Permission, capability: Capability)
-                                    (applicationId: String)
-                                    (fun: ApplicationRequest[AnyContent] => Future[Result]): Action[AnyContent] = {
-    loggedInAction { implicit request =>
-      val composedActions = Action andThen applicationAction(applicationId, loggedIn) andThen permissionFilter(permissions) andThen capabilityFilter(capability)
-      composedActions.async(fun)(request)
-    }
-  }
+  def capabilityThenPermissionsActionForAllStates = capabilityThenPermissionsActionWithStateCheck(stateCheck = _ => true) _
```
Comment (adampridmore) [inline app/controllers/Controller.scala:123]: checkActionForXXX name

File: app/controllers/Controller.scala [hunk 3]
```diff
@@ -86,6 +86,12 @@ abstract class ApplicationController()
 
   implicit def userFromRequest(implicit request: ApplicationRequest[_]): DeveloperSession = request.user
 
+  def applicationStateApprovedPredicate(state: State) = state.isApproved
+
+  def applicationStateTesting(state: State) = state.isInTesting
+
+  def applicationStateApprovedOrTestingPredicate(state: State) = state.isApproved || state.isInTesting
+
```
Comment (AndySpaven) [inline app/controllers/Controller.scala:94]: Inline these

File: app/controllers/Controller.scala [hunk 4]
```diff
@@ -100,30 +106,34 @@ abstract class ApplicationController()
       composedActions.async(fun)(request)
     }
 
-  def capabilityThenPermissionsAction(capability: Capability, permissions: Permission)
-                                     (applicationId: String)
-                                     (fun: ApplicationRequest[AnyContent] => Future[Result]): Action[AnyContent] = {
+  private def capabilityThenPermissionsActionWithStateCheck(stateCheck: State => Boolean)(capability: Capability, permissions: Permission)
```
Comment (AndySpaven) [inline app/controllers/Controller.scala:109]: Shorten name to actionWithStateCheck or similar

File: app/controllers/Controller.scala [hunk 5]
```diff
@@ -100,30 +106,34 @@ abstract class ApplicationController()
...
-                                    (applicationId: String)
-                                    (fun: ApplicationRequest[AnyContent] => Future[Result]): Action[AnyContent] = {
-    loggedInAction { implicit request =>
-      val composedActions = Action andThen applicationAction(applicationId, loggedIn) andThen permissionFilter(permissions) andThen capabilityFilter(capability)
-      composedActions.async(fun)(request)
-    }
-  }
+  def capabilityThenPermissionsActionForAllStates = capabilityThenPermissionsActionWithStateCheck(stateCheck = _ => true) _
+
+  def capabilityThenPermissionsActionForApprovedApps = capabilityThenPermissionsActionWithStateCheck(stateCheck = applicationStateApprovedPredicate) _
```
Comment (AndySpaven) [inline app/controllers/Controller.scala:125]: Following inline above

`...WithStateCheck( (state: State) => state.isApproved )
`
or similar

File: app/domain/Capabilities.scala [hunk 6]
```diff
@@ -43,17 +38,17 @@ object Capabilities {
   }
 
   case object ChangeClientSecret extends Capability {
-    def hasCapability(app: Application) = app.state.name == State.PRODUCTION
+    def hasCapability(app: Application) = app.state.name.isApproved
```
Comment (AndySpaven) [inline app/domain/Capabilities.scala:41]: Is it time to add helper methods to app so we can do
`app.isApproved`

and not care how this is done.

Likewise for other tests below.

**Pull request merged**